pop-api/examples/nonfungibles/tests.rs

```diff
@@ -0,0 +1,485 @@
+		ALICE.into(),
+		None
+	));
+	// Fails with `TokenNotExists`.
+	assert_eq!(contract.burn(&mut session, ITEM + 1), Err(Psp34Error::TokenNotExists));
+}
+
+#[drink::test(sandbox = Pop)]
+fn transfer_works(mut session: Session) {
+	let _ = env_logger::try_init();
+	// Deploys a new contract.
+	let contract = Contract::default(&mut session).unwrap();
+	assert_ok!(session.sandbox().mint(
+		Some(contract.address.clone()),
+		COLLECTION,
+		ITEM,
+		contract.address.clone().into(),
+		None
+	));
+	// Successfully transfers an item.
+	assert_ok!(contract.transfer(&mut session, ALICE, ITEM));
+	assert_eq!(session.sandbox().owner(&COLLECTION, &ITEM), Some(ALICE));
+	assert_eq!(session.sandbox().balance_of(&COLLECTION, &ALICE), 1);
+	assert_last_contract_event!(
+		&session,
+		Transfer {
+			from: Some(account_id_from_slice(&contract.address)),
+			to: Some(account_id_from_slice(&ALICE)),
+			item: ITEM
+		}
+	);
+}
+
+#[drink::test(sandbox = Pop)]
+fn transfer_fails_with_unauthorization(mut session: Session) {
+	let _ = env_logger::try_init();
+	// Deploys a new contract.
+	let contract = Contract::default(&mut session).unwrap();
+	assert_ok!(session.sandbox().mint(
+		Some(contract.address.clone()),
+		COLLECTION,
+		ITEM,
+		contract.address.clone().into(),
+		None
+	));
+	// Fails with `Not the contract owner`.
+	session.set_actor(BOB);
+	assert_eq!(
+		contract.transfer(&mut session, ALICE, ITEM),
+		Err(Psp34Error::Custom("Not the contract owner".to_string()))
+	);
+}
+
+#[drink::test(sandbox = Pop)]
+fn transfer_fails_with_not_approved(mut session: Session) {
+	let _ = env_logger::try_init();
+	// Deploys a new contract.
+	let contract = Contract::default(&mut session).unwrap();
+	assert_ok!(session.sandbox().mint(
+		Some(contract.address.clone()),
+		COLLECTION,
+		ITEM,
+		ALICE.into(),
+		None
+	));
+	// Fails with `NotApproved`.
+	assert_eq!(contract.transfer(&mut session, ALICE, ITEM), Err(Psp34Error::NotApproved));
+}
+
+#[drink::test(sandbox = Pop)]
+fn transfer_fails_with_invalid_item(mut session: Session) {
+	let _ = env_logger::try_init();
+	// Deploys a new contract.
+	let contract = Contract::default(&mut session).unwrap();
+	assert_ok!(session.sandbox().mint(
+		Some(contract.address.clone()),
+		COLLECTION,
+		ITEM,
+		ALICE.into(),
+		None
+	));
+	// Fails with `TokenNotExists`.
+	assert_eq!(contract.transfer(&mut session, ALICE, ITEM + 1), Err(Psp34Error::TokenNotExists));
+}
+
+#[drink::test(sandbox = Pop)]
+fn destroy_works(mut session: Session) {
+	let _ = env_logger::try_init();
+	// Deploys a new contract.
+	let contract = Contract::default(&mut session).unwrap();
+	// Successfully destroys a collection.
+	session.set_gas_limit(Weight::MAX);
+	let witness_string =
+		format!("{:?}", DestroyWitness { item_metadatas: 0, item_configs: 0, attributes: 0 });
+
+	// Error returned "Not enough data to fill buffer" due to contract termination.
+	assert!(session.call::<String, ()>("destroy", &[witness_string], None).is_err());
```

Comment:
This could be improved imo. it seems to stem from the fact that the call impl in pop-drink panics, eliminating the possibility of handling this cleanly. This means that every contract will have this messy approach to termination.

The "Not enough data to fill buffer" should not be leaked out beyond pop-drink - perhaps it can be converted to a known error type?